MVC Step3 띠용

jye326 opened the pull request:
이번에는 요구사항만 만족하도록 빠르게 구현해보았습니다. 이번에도 멋진 피드백 부탁드려요~

jye326 committed:
test: 리플렉션 테스트 수정 및 누락된 메소드 구현

- 클래스 네이밍 테스트 수정
- 필드 및 메소드 추출 로직 구현
- 생성자 관련 테스트 케이스 수정
- Junit3/4TestRunner 실행 로직 구현
- 필요한 테스트 라이브러리 추가

jye326 committed:
feat: 추가 Servlet 구현 및 CharacterEncodingFilter 수정

- CharacterEncodingFilter: response content type 설정 추가
- FormHandlingServlet: GET/POST 요청 처리 로직 구현
- HttpMethodServlet: HTTP 메서드 처리 서비스 구현
- BasicServlet: 기본 Servlet 동작 확인용 구현

jye326 committed:
fix: enable filter chain execution in CharacterEncodingFilter

jye326 committed:
build: JUnit 의존성 구성 변경 및 platform BOM 추가

jye326 committed:
feat: HandlerExecution 클래스에 동적 메서드 호출 추가

- 동적 메서드 호출을 위한 handler와 method 필드 추가
- 메서드 접근 허용을 위해 setAccessible(true) 호출
- HttpServletRequest와 HttpServletResponse를 인자로 받는 메서드 호출 로직 구현

jye326 committed:
feat: AnnotationHandlerMapping 클래스에 어노테이션 기반 핸들러 매핑 추가

- @Controller와 @RequestMapping 어노테이션 기반으로 핸들러 자동 스캔 구현
- HandlerKey와 HandlerExecution을 활용한 핸들러 등록 로직 추가
- HTTP 요청 메서드와 URI 매핑을 통한 핸들러 검색 기능 구현
- URI 정규화 메서드 추가

jye326 committed:
feat: JspView 클래스 구현 및 DispatcherServlet에 적용

- JspView에서 JSP 뷰 렌더링 및 리다이렉트 처리 로직 추가
- DispatcherServlet에 JspView를 이용한 뷰 렌더링 로직 수정
- 불필요한 move 메서드 제거 및 관련 뷰 처리 로직 간소화

jye326 committed:
refactor: AnnotationHandlerMapping 메서드 분리 및 코드 가독성 개선

- initialize 메서드 내부 로직 handlerMapping 메서드로 분리
- stream 사용으로 메서드 매핑 로직 간결화
- 로깅 추가 및 코드 구조 정리

jye326 committed:
refactor: JspView 리다이렉트 처리 로직 분리 및 ViewResolver 도입

- JspView 내부 리다이렉트 처리 로직 ViewResolver로 이동
- DispatcherServlet에서 ViewResolver를 활용한 리다이렉트 처리 추가
- 코드 구조 개선 및 가독성 향상
- 주석 수정 및 필요 없는 import 정리

jye326 committed:
refactor: Controller 인터페이스 리팩토링 및 반환 타입 수정

- Controller 인터페이스 execute 메서드 반환 타입 변경 (String → ModelAndView)
- View 반환 로직에서 JspView 객체로 전환
- 가독성 향상을 위한 코드 최적화 및 로깅 추가

jye326 committed:
refactor: ControllerScanner 도입 및 AnnotationHandlerMapping 리팩토링

- @Controller 스캔 및 인스턴스 매핑 역할 ControllerScanner로 분리
- initialize 메서드 리팩토링 및 매핑 로직 간소화
- ReflectionUtils 활용으로 메서드 검색 및 핸들러 매핑 코드 최적화
- 코드 구조 개선 및 로깅 수정

jye326 committed:
refactor: HandlerAdapter 도입 및 DispatcherServlet 리팩토링

- 핸들러 처리를 위한 HandlerAdapter 클래스 추가
- DispatcherServlet 일부 로직 HandlerAdapter로 위임
- AnnotationHandlerMapping 초기화 로직 및 핸들러 등록 추가
- 예외 상황을 처리하기 위한 NoHandlerFoundException 클래스 추가
- JspView 리다이렉트 처리 조건 추가 및 코드 정리

jye326 committed:
Merge branch 'step2' of https://github.com/jye326/java-mvc into step2

# Conflicts:
#	app/src/main/java/com/techcourse/DispatcherServlet.java
#	mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java
#	mvc/src/main/java/com/interface21/webmvc/servlet/view/JspView.java

jye326 committed:
refactor: ModelAndView 객체 생성 로직 간소화

- ModelAndView.ofJspView 정적 메서드 추가
- 기존 new JspView 인스턴스 생성 코드 변경 및 중복 제거
- 불필요한 import문 정리

jye326 committed:
refactor: ControllerScanner 클래스 주석 개선

- 인스턴스 매핑 관련 주석 내용 명확화

jye326 committed:
refactor: JspView 및 AnnotationHandlerMapping 주석 개선

- JspView 클래스의 주석 줄바꿈 수정
- AnnotationHandlerMapping 클래스에 initialize 메서드 관련 주석 추가 및 리플렉션 기반 매핑 로직 설명 보완

jye326 committed:
refactor: HandlerMapping 인터페이스에 initialize 메서드 추가 및 ManualHandlerMapping 구현 수정

- HandlerMapping 인터페이스에 initialize 메서드 정의
- ManualHandlerMapping 클래스에 initialize 메서드 구현 및 Override 추가

jye326 committed:
refactor: DispatcherServlet 및 HandlerAdapter 코드 간소화

- handlerMappings 초기화 방식 변경 및 반복 초기화 제거
- HandlerAdapter 메서드 시그니처에서 HandlerMapping 의존성 제거
- 적절한 핸들러를 찾지 못한 경우 404 에러 응답 처리 추가
- service 메서드 내 핸들러 매핑 탐색 로직 명확화 및 주석 추가

jye326 committed:
refactor: DispatcherServlet 핸들러 검색 로직 간소화

- 불필요한 handler null 초기화 코드 제거
- 핸들러 매핑에서 적절한 핸들러 발견 시 즉시 반환하도록 수정
- 404 응답 처리 로직 간결화

jye326 committed:
refactor: UriUtils 유틸 클래스 추가 및 URI 처리 로직 개선

- ManualHandlerMapping에서 URI 처리 로직 UriUtils로 이동
- ResourceFilter의 URI 처리 코드 개선 및 UriUtils 활용
- URI 처리 전용 유틸리티 메서드 getResourcePath 추가

jye326 committed:
refactor: ControllerScanner의 getControllers 메서드 반환값 수정

- 반환 시 새로운 HashMap 인스턴스를 생성하여 불변성 확보

jye326 committed:
refactor: HandlerAdapter의 render 메서드 호출 시 Model 전달 방식 수정

- Map.of() 대신 ModelAndView의 getModel로 모델 객체 전달
- 불필요한 Map import 제거

jye326 committed:
refactor: DispatcherServlet 핸들러 매핑 초기화 순서 수정

- ManualHandlerMapping 초기화 순서를 AnnotationHandlerMapping 이후로 조정

jye326 committed:
refactor: AnnotationHandlerMapping 패키지 경로 수정

- 파일 경로 구분자를 "/"에서 "."으로 변경하여 표준 패키지 표현 방식 사용

jye326 committed:
refactor: support 메서드 추가 및 DispatcherServlet 처리 로직 개선

- HandlerMapping 인터페이스에 support 메서드 추가
- AnnotationHandlerMapping 및 ManualHandlerMapping에 support 메서드 구현
- DispatcherServlet에서 핸들러 처리 조건에 support 메서드 적용
- 코드 가독성 향상을 위한 로직 단순화 및 리팩토링

jye326 committed:
refactor: HandlerAdapter 인터페이스 분리 및 DispatcherServlet 리팩토링

- HandlerAdapter 인터페이스로 리팩토링 및 클래스 분리
- ControllerAdapter, HandlerExecutionAdapter 구현체 추가
- DispatcherServlet에서 handler 처리 로직을 handlerAdapters로 위임
- 코드 구조 개선 및 가독성 향상

jye326 committed:
refactor: 예외 클래스 리팩토링 및 패키지 이동

- NoHandlerFoundException 등 예외 클래스 패키지를 exception으로 이동
- NoHandlerAdapterFoundException 및 NoHandlerMappingFoundException 클래스 추가
- 예외 처리 구분을 위한 클래스 분리 및 가독성 향상

jye326 committed:
refactor: ControllerScanner 필드명 변경 및 AnnotationHandlerMapping 코드 가독성 개선

- controllers 필드명을 controllerRegistry로 변경
- 관련 메서드 및 변수명 일괄 수정
- AnnotationHandlerMapping의 initialize 메서드 내 코드 간결화
- 불필요한 코드 제거 및 가독성 향상

jye326 committed:
refactor: AnnotationHandlerMapping 메서드 시그니처 변경 및 코드 개선

- mappingAllHandlerMethods 매개변수 수정 (class → entry 활용)
- 메서드 호출 로직 간결화 및 코드 가독성 향상
- 불필요한 변수 제거 및 개선된 로직 적용

jye326 committed:
refactor: DispatcherServlet 및 불필요한 HandlerAdapterrrr 제거

- HandlerMapping 및 HandlerAdapter 탐색 로직을 메서드로 분리하여 가독성 향상
- 사용하지 않는 HandlerAdapterrrr 클래스 삭제
- 404 처리 시 응답 코드 전송 및 예외 메시지 명확화

jye326 committed:
refactor: HandlerAdapter의 handle 메서드 반환값 수정 및 DispatcherServlet 로직 개선

- HandlerAdapter의 handle 메서드 반환 타입을 ModelAndView로 변경
- DispatcherServlet에서 ModelAndView 반환 처리 및 View 렌더링 로직 추가
- ModelAndView의 모델 객체 활용을 위한 Map 타입 import 추가

jye326 committed:
refactor: AnnotationHandlerMapping의 URI 처리 로직 개선 및 UriUtils 추가

- URI 처리 코드를 UriUtils.getResourcePath 메서드로 분리
- AnnotationHandlerMapping에서 UriUtils 활용하여 요청 URI 처리

jye326 committed:
refactor: DispatcherServlet의 예외 처리 및 404 응답 로직 개선

- NoHandlerMappingFoundException 발생 시 404 응답 코드 전송 처리 추가
- service 메서드의 throws 선언에 IOException 추가
- 기존 404 응답 처리 중복 코드 제거

jye326 committed:
refactor: AnnotationHandlerMapping의 URI 정규화 로직 수정 및 UriUtils 개선

- URI 정규화를 UriUtils.normalize 메서드로 통합
- AnnotationHandlerMapping에서 요청 URI 처리 시 정규화 적용
- UriUtils에 URI 정규화 메서드 추가 및 getResourcePath 활용 방식 개선

jye326 committed:
feat: UserController 추가 및 기본 API 테스트 작성

- GET /api/user 엔드포인트 처리하는 UserController 추가
- InMemoryUserRepository 활용하여 사용자 조회 기능 구현
- JsonView를 사용하여 사용자 정보를 JSON으로 반환
- UserControllerTest 작성 및 show 메서드 동작 검증

jye326 committed:
feat: JsonView에 JSON 직렬화 및 응답 처리 로직 추가

- ObjectMapper를 활용한 JSON 직렬화 로직 추가
- 응답 Content-Type을 "application/json;charset=UTF-8"로 설정
- 모델 값이 1개일 경우 해당 값만 JSON으로 직렬화
- PrintWriter를 사용해 JSON 데이터를 HTTP 응답에 작성 및 전송 추가

jye326 committed:
refactor: asis 패키지의 기존 MVC 컴포넌트 제거 및 어노테이션 기반 리팩토링

- asis 패키지의 Controller, ControllerAdapter, ManualHandlerMapping 클래스 제거
- 어노테이션 기반의 DispatcherServlet 및 관련 구성 리팩토링
- 각 Controller 클래스에 @Controller 및 @RequestMapping 어노테이션 추가

jye326 committed:
merge

2Jin1031 commented on app/src/main/java/com/techcourse/controller/ForwardController.java:
```diff
@@ -0,0 +1,17 @@
+@Controller
+public class ForwardController {
+
+    @RequestMapping(value = "/", method = RequestMethod.GET)
+    public ModelAndView execute(final HttpServletRequest request, final HttpServletResponse response) {
+        return ModelAndView.ofJspView("/index.jsp");
+    }
```
컨트롤러의 반환 형태가 ModelAndView로 처리되어 있어서
내부적으로 JspView로 변환하는 로직이 포함되어 있는 것 같아요
컨트롤러는 단순히 실행 결과만 반환하고, 반환값을 해석해 처리하는 과정은`HandlerExecution` 의 역할로 두는 건 어떻게 생각하는 지 궁금해요!

jye326 replied:
반환 타입을 Object로 변경하고 `HandlerExecution`에서 타입 별로 처리하는 형태를 말씀하신 것이 맞나요? 그것도 객체의 책임 분리 측면에서 좋은 방법이라고 생각해요! 그런데 컨트롤러의 반환 타입을 ModelAndView로 제한하는 것도 프레임 워크로서의 의미가 있다고 생각합니다!

2Jin1031 replied:
MVC 프레임워크를 직접 만들고 학습하는 과정에서, 컨트롤러의 반환 타입을 제한하는 것이 곧 프레임워크의 의미와 연결된다고 보는 것이 저로서는 잘 와닿지 않는 것 같아요 ㅜ
띠용이 말씀해주신 맥락을 그대로 적용한다면, 오히려 Controller의 책임을 좁히고 해석 책임을 분리하는 것이 프레임워크의 학습의 핵심 포인트로 볼 수도 있지 않을까요? 🧐
실제 프레임워크는 반환 타입을 제한하지 않고, 다양한 반환값을 지원하면서도 HandlerAdapter 등이 책임을 맡고 있어서요!

2Jin1031 reviewed (changes requested):
step3까지 고생 많았어요 띠용~

Legacy 컨트롤러 코드를 어노테이션으로 수정한 거 확인했습니다
요구사항을 충분히 만족한 것 같아 따로 코멘트 달 부분이 보이지 않았던 것 같아요
(구현하면서 궁금하신 점이 있으셨다면 언제든지 추가 요청 주세요)

다만, 몇몇 페이지들이 정상적으로 동작하지 않는 것 같습니다
아마 `DispatcherServletInitializer` 위치 변경이 원인일 수 있으니
다시 한번 확인해주시면 될 것 같습니다

다음번에 approve & merge 진행할게요~

jye326 committed:
feat: RegisterController 메서드 경로 수정 및 Gradle 태스크 추가

- RegisterController의 @RequestMapping 경로를 "/logout"에서 "/register"로 수정
- Gradle에 copyMvcClassesToWebInf 태스크 추가 및 classes 태스크 의존성으로 설정

jye326 commented on app/build.gradle:
```diff
@@ -45,3 +49,12 @@ sourceSets {
     }
 }
+
+tasks.register('copyMvcClassesToWebInf', Copy) {
+    from(project(':mvc').sourceSets.main.output)
+    into("$projectDir/src/main/webapp/WEB-INF/classes")
+}
+
+tasks.named('classes') {
+    dependsOn 'copyMvcClassesToWebInf'
+}
```
mvc 모듈을 따로 빌드한 뒤, 해당 산출물들을 현재 프로젝트(app)의 톰캣이 인식하는 `WEB-INF/classes` 로 복사합니다. 이거 안해주면 `DispatcherServletInitializer` 인식을 못하더라고요.

2Jin1031 replied:
작동이 잘 되는 거 확인했습니다 ~
너무 좋네요 🤙

2Jin1031 reviewed (approved):
step3까지 넘 고생 많으셨어요~ 띠용
리뷰 코멘트를 남기는 과정에서 덕분에 많이 배울 수 있었습니다~
JDBC 미션도 화이팅입니다! 🔥🔥🔥
